x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/SyntheticSourceLicenseService.java

```diff
@@ -30,39 +40,80 @@ final class SyntheticSourceLicenseService {
         Setting.Property.Dynamic
     );
 
-    private static final LicensedFeature.Momentary SYNTHETIC_SOURCE_FEATURE = LicensedFeature.momentary(
+    static final LicensedFeature.Momentary SYNTHETIC_SOURCE_FEATURE = LicensedFeature.momentary(
         MAPPINGS_FEATURE_FAMILY,
         "synthetic-source",
         License.OperationMode.ENTERPRISE
     );
 
+    static final LicensedFeature.Momentary SYNTHETIC_SOURCE_FEATURE_LEGACY = LicensedFeature.momentary(
+        MAPPINGS_FEATURE_FAMILY,
+        "synthetic-source-legacy",
+        License.OperationMode.GOLD
+    );
+
+    private final long cutoffDate;
+    private LicenseService licenseService;
     private XPackLicenseState licenseState;
     private volatile boolean syntheticSourceFallback;
 
     SyntheticSourceLicenseService(Settings settings) {
-        syntheticSourceFallback = FALLBACK_SETTING.get(settings);
+        this(settings, System.getProperty(CUTOFF_DATE_SYS_PROP_NAME));
+    }
+
+    SyntheticSourceLicenseService(Settings settings, String cutoffDate) {
+        this.syntheticSourceFallback = FALLBACK_SETTING.get(settings);
+        this.cutoffDate = getCutoffDate(cutoffDate);
     }
 
     /**
      * @return whether synthetic source mode should fallback to stored source.
      */
-    public boolean fallbackToStoredSource(boolean isTemplateValidation) {
+    public boolean fallbackToStoredSource(boolean isTemplateValidation, boolean legacyLicensedUsageOfSyntheticSourceAllowed) {
         if (syntheticSourceFallback) {
             return true;
         }
 
+        var operationMode = licenseState.getOperationMode();
```

Comment:
To avoid potential races on license state changing between multiple reads, lets use a copy here, i.e.,:

`XPackLicenseState licenseStateSnapshot = licenseState.copyCurrentLicenseState();`

Then use `licenseStateSnapshot` in the code below